toolchain/driver/driver.cpp

```diff
@@ -22,286 +25,481 @@
-  DiagnosticConsumer* consumer = &stream_consumer;
-  std::unique_ptr<SortingDiagnosticConsumer> sorting_consumer;
-  // TODO: Figure out a command-line support library, this is temporary.
-  if (!args.empty() && args[0] == "-v") {
-    args = args.drop_front();
-    // Note this implies streamed output in order to interleave.
-    vlog_stream_ = &error_stream_;
-  } else if (!args.empty() && args[0] == "--print-errors=streamed") {
-    args = args.drop_front();
-  } else {
-    sorting_consumer = std::make_unique<SortingDiagnosticConsumer>(*consumer);
-    consumer = sorting_consumer.get();
+  enum class Phase {
+    Tokenize,
+    Parse,
+    Syntax,
+    Lower,
+    CodeGen,
+  };
+
+  friend auto operator<<(llvm::raw_ostream& out, Phase phase)
+      -> llvm::raw_ostream& {
+    switch (phase) {
+      case Phase::Tokenize:
+        out << "tokenize";
+        break;
+      case Phase::Parse:
+        out << "parse";
+        break;
+      case Phase::Syntax:
+        out << "syntax";
+        break;
+      case Phase::Lower:
+        out << "lower";
+        break;
+      case Phase::CodeGen:
+        out << "codegen";
+        break;
+    }
+    return out;
   }
 
-  if (args.empty()) {
-    error_stream_ << "ERROR: No subcommand specified.\n";
-    return false;
+  void Build(Args::CommandBuilder& b) {
+    b.AddStringPositionalArg(
+        {
+            .name = "FILE",
+            .help = R"""(
+The input Carbon source file to compile.
+)""",
+        },
+        [&](auto& arg_b) {
+          arg_b.Required(true);
+          arg_b.Set(&input_file_name);
+        });
+
+    b.AddOneOfOption(
+        {
+            .name = "phase",
+            .help = R"""(
+Selects the compilation phase to run. These phases are always run in sequence,
+so every phase before the one selected will also be run. The default is to
+compile, lower, and generate machine code.
+)""",
+        },
+        [&](auto& arg_b) {
+          arg_b.SetOneOf(
+              {
+                  arg_b.OneOfValue("tokenize", Phase::Tokenize),
+                  arg_b.OneOfValue("parse", Phase::Parse),
+                  arg_b.OneOfValue("syntax", Phase::Syntax),
+                  arg_b.OneOfValue("lower", Phase::Lower),
+                  arg_b.OneOfValue("codegen", Phase::CodeGen).Default(true),
+              },
+              &phase);
+        });
+
+    b.AddStringOption(
+        {
+            .name = "output",
+            .value_name = "FILE",
+            .help = R"""(
+The output filename for codegen.
+
+When this is a file name, either textual assembly or a binary object will be
+written to it based on the flag `--asm-output`. The default is to write a binary
+object file.
+
+This can also be the string `-` which will write the output to stdout. In that
+case, the flag `--asm-output` is ignored and the output defaults to textual
+assembly. Binary object output can be forced by enabling `--force-obj-output`.
+)""",
```

Comment:
What is the purpose of allowing '-' as an output file name?  If the user doesn't provide an output file name, based on the codegen flag we can emit the assembly to stdout. However, if the codegen flag is set to --force-obj-output and the output file is not given, we can either throw an error or emit the object code to a default output file.

Reply:
This is a somewhat common convention for directing output to stdout instead of a file:

```
> clang -c -S -o - example.c
        .section        __TEXT,__text,regular,pure_instructions
        .build_version macos, 13, 0
        .globl  _main                           ; -- Begin function main
        .p2align        2
_main:                                  ; @main
        .cfi_startproc
; %bb.0:
        sub     sp, sp, #16
        .cfi_def_cfa_offset 16
        mov     w0, #0
        str     wzr, [sp, #12]
        add     sp, sp, #16
        ret
        .cfi_endproc
                                        ; -- End function
.subsections_via_symbols
```

The common convention in compilers (Clang and GCC at least) is to select an output name based on the input name when an output flag isn't provided:

```
> clang -c -S example.c
> cat example.s
        .section        __TEXT,__text,regular,pure_instructions
        .build_version macos, 13, 0
        .globl  _main                           ; -- Begin function main
        .p2align        2
_main:                                  ; @main
        .cfi_startproc
; %bb.0:
        sub     sp, sp, #16
        .cfi_def_cfa_offset 16
        mov     w0, #0
        str     wzr, [sp, #12]
        add     sp, sp, #16
        ret
        .cfi_endproc
                                        ; -- End function
.subsections_via_symbols
```

While for testing, it is nice to have this streamed to stdout, that's somewhat rarely the desired behavior.

My goal with all of the driver changes like this is to try to line up the behavior with what is more conventional for compilers on the command line.

I'll try to write up a summary of this in the up-coming split PR description as well.